src/TorchSharp/DisposeScopeManager.cs

```diff
@@ -0,0 +1,317 @@
+                return (first, second, third);
+            }
+
+            /// <summary>
+            /// Disposes everything currenly in the dispose scope.
+            /// </summary>
+            public void DisposeEverything() => DisposeEverythingBut();
+
+            /// <summary>
+            /// As an intermediate step, you can dispose all the tensors/disposables currently scheduled for dispose, to
+            /// clear up some memory without creating a new scope. Note that this doesn't permanently exclude the
+            /// tensors from disposing, use Exclude for that. Also, excluded tensors don't need to be included
+            /// here.
+            /// </summary>
+            public IDisposable[] DisposeEverythingBut(params IDisposable[] keep)
+            {
+                lock (this) {
+                    foreach (var disposable in keep) {
+                        if (!Disposables.Contains(disposable)) {
+                            throw new InvalidOperationException("The disposable does not belong to this scope!");
+                        }
+                    }
+
+                    var toDispose = Disposables.Where(disposable => !keep.Contains(disposable)).ToList();
+                    foreach (var disposable in toDispose) {
+                        DoDispose(disposable);
+                    }
+                }
+
+                return keep;
+            }
+
+            public T DisposeEverythingBut<T>(T keep) where T : IDisposable
+            {
+                DisposeEverythingBut(new IDisposable[] { keep });
+                return keep;
+            }
+
+            public (T1 first, T2 second) DisposeEverythingBut<T1, T2>(T1 first, T2 second)
+                where T1 : IDisposable where T2 : IDisposable
+            {
+                DisposeEverythingBut(new IDisposable[] { first, second });
+                return (first, second);
+            }
+
+            public (T1 first, T2 second, T3 third) DisposeEverythingBut<T1, T2, T3>(T1 first, T2 second, T3 third)
+                where T1 : IDisposable where T2 : IDisposable where T3 : IDisposable
+            {
+                DisposeEverythingBut(new IDisposable[] { first, second, third });
+                return (first, second, third);
+            }
+
+            public void Dispose()
+            {
+                ReleaseUnmanagedResources();
+                GC.SuppressFinalize(this);
+            }
+
+            private void ReleaseUnmanagedResources()
+            {
+                lock (Singleton) {
+                    foreach (var disposable in Disposables) {
+                        DoDispose(disposable);
+                    }
+
+                    Singleton.RemoveDisposeScope(this);
+                }
+            }
+
+            private void DoDispose(IDisposable disposable)
+            {
+                _disposeScopeManager.CurrentlyDisposing = disposable;
+                if (disposable is torch.Tensor tensor) {
```

Comment:
How about a try ... finally here to make sure that the 'CurrentlyDisposing' is always set to null after? I'm not sure that it's necessary, but it feels like one of those situations where I would expect it.

Reply:
Fixed